File: ruby/circular-buffer/circular_buffer.rb
class CircularBuffer

  class BufferEmptyException < StandardError ; end
  class BufferFullException < StandardError ; end

  def initialize(size)
    @size = size
    clear
  end

  def clear
    @data = Array.new(@size)
    @read_pt_index, @write_pt_index = 0, 0
  end

  def read
    raise BufferEmptyException if @data.compact.empty?
    data = @data[@read_pt_index]
    @data[@read_pt_index] = nil
    move_pt_index(:reader)
    data
  end

  def write!(element)
    write(element, true)
  end

  def write(element, override=false)
    # when data is not empty, reader pointer meets writer pointer and in non-overrride mode
    raise BufferFullException if !override && @read_pt_index == @write_pt_index && !@data.compact.empty?
    @data[@write_pt_index] = element
    # move reader pointer in override mode
    move_pt_index(:reader) if override && @read_pt_index == @write_pt_index
    move_pt_index(:writer)
    self
  end

  private

  def move_pt_index(pt_type)
    pt_type == :reader ? @read_pt_index = (@read_pt_index + 1) % @size
                       : @write_pt_index = (@write_pt_index + 1) % @size
  end

end

File: python/circular-buffer/circular_buffer.py
class BufferFullException(Exception):
    pass


class BufferEmptyException(Exception):
    pass


class CircularBuffer(object):
    def __init__(self, capacity):
        self.lst = [None for i in range(capacity)]
        self.start_index = capacity // 2
        self.end_index = capacity // 2

    def read(self):
        if self.start_index == self.end_index and all(x == None for x in self.lst):
            raise BufferEmptyException('Empty')
        
        data = self.lst[self.start_index]
        self.lst[self.start_index] = None
        self.start_index = (self.start_index + 1) % len(self.lst)
        return data
 

    def write(self, data):
        if self.start_index == self.end_index and all(x != None for x in self.lst):
            raise BufferFullException('Full')
        self.overwrite(data)

    def overwrite(self, data):
        self.lst[self.end_index] = data
        if self.start_index == self.end_index and all(x != None for x in self.lst):
            self.start_index = (self.start_index + 1) % len(self.lst)
        self.end_index = (self.end_index + 1) % len(self.lst)

    def clear(self):
        self.lst = [None for i in range(len(self.lst))]
